src/development/metrics.md

```diff
@@ -0,0 +1,153 @@
+
+### Witnesses
+
+#### Size
+
+Research has been conducted, in order to determine the best structure for a witness. We tried three methods:
+
+ * The first one, as specified in the [current version of the consensus spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip6800/beacon-chain.md#executionwitness), referred to as **type 0**. This approach uses the SSZ type `Optional[T]` , that is yet to be officially included in the spec.
+ * A first variation of the spec, in which prestate values are grouped together in their own list, and poststate values are grouped in their own list as well. This is referred to as **type 1**.
+
+```python
+class SuffixStateDiffs(Container):
+    suffixes: bytes
+    # None means not currently present
+    current_values: List[bytes]
+    # None means value not updated
+    new_values:     List[bytes]
+```
+Type 1 serialization container
+
+ * A second variation of the spec, in which updates, insertions and reads are grouped in their own separate lists. The suffixes for each of these lists are also grouped as their own byte lists. This is referred to as **type 2**.
+
+```python
+class SuffixStateDiffs(Container):
+    updated_suffixes: bytes
+    updated_currents: List[Bytes32]
+    updated_news: List[Bytes32]
+
+    inserted_suffixes: bytes
+    inserted_news: List[Bytes32]
+
+    read_suffixes: bytes
+    read_currents: List[Bytes32]
+
+    missing_suffixes: bytes
+```
+Type 2 serialization container
+
+ * A final variation, where update, reads, inserts and proof-of-absences are grouped, along with their suffixes, into their own lists. This is referred to as **type 3**.
+
+```python
+class UpdateDiff(Container):
+    suffix: byte
+    current: Bytes32
+    new: Bytes32
+
+class ReadDiff(Container):
+    suffix: byte
+    current: Bytes32
+
+class InsertDiff(Container):
+    suffix: byte
+    new: Bytes32
+
+class MissingDiff(Container):
+    suffix byte
+
+class StemStateDiff(Container):
+    stem: Bytes31
+
+    updates: List[UpdateDiff]
+    reads: List[ReadStateDiff]
+    insert: List[InsertStateDiff]
+    # TODO: check if this encodes as well as bytes
+    missing: List[MissingStateDiff]
+```
+Type 3 serialization container
+
+Here are our findings after replaying data for 200k blocks:
+
+![image.png](./assets/total.png)
+
+One can see that the spec is much less efficient than the other types. This is due to the inefficiency of storing `Optional[T]`. As a result, we didn’t include this method in the rest of the analysis.
+
+When only considering the post-transition witnesses, this is what is found:
+
+![image.png](./assets/compare_post_transition.png)
+
+The maximum and minimum sizes for each type, are summarized in the following table:
```

Comment:
Uhm, the "minimum" doesn't seem to be the same opposite definition as "maximum", right?
Should say "minimum (without outliers)"? Since the minimum of all cases is close to 0?

Reply:
clarified to "minimum observed"